Add extension guides to terraform provider (#590)
nitrictech/docs: src/pages/guides/terraform/api-gateway-throttle.mdx

export const description =

>   'Extend the Nitric API Gateway Terraform provider to set throttling limits'

Reviewer: ```suggestion
  'Extend the Nitric AWS Terraform provider to set API Gateway throttling limits'
```

# Add Throttle Limits to API Gateway

> Throttling helps in preventing your backend services from being overwhelmed by too many requests at once. This is particularly important if your backend services have limited capacity and can only handle a certain number of requests per second.
> 
> Without throttling, a single client could potentially consume all available resources, leaving others with degraded service or no service at all.

Reviewer: ```suggestion
Throttling helps prevent your backend services from being overwhelmed by too many requests at once. This is particularly important if your backend services have limited capacity and can only handle a certain number of requests per second.

Without throttling, a single client could potentially consume all available resources, leaving others with degraded service or no service at all.
```

## What we'll be doing

> 1. Review our existing module

Reviewer: ```suggestion
1. Review the existing module
```

2. Configure throttle limits

> ## Review our existing module

Reviewer: ```suggestion
## Review the existing module
```

> The first step is clone the [Nitric repository](https://github.com/nitrictech/nitric) and examine how the [Nitric Terrraform provider provisions an API Gateway](https://github.com/nitrictech/nitric/tree/main/cloud/aws/deploytf/.nitric/modules/api).

Reviewer: ```suggestion
Start by cloning the [Nitric repository](https://github.com/nitrictech/nitric, then examine how the Terrraform provider [provisions an API Gateway](https://github.com/nitrictech/nitric/tree/main/cloud/aws/deploytf/.nitric/modules/api).
```

```bash
git clone https://github.com/nitrictech/nitric
cd nitric
```

The AWS API module in the default Terraform provider performs the following tasks:

1. Defines an HTTP API Gateway with specified name, protocol, and API specification.
2. Sets up a "$default" deployment stage with automatic deployment enabled.
3. Allows the API Gateway to invoke specified Lambda functions.
4. Looks up existing SSL certificates for specified domains.
5. Configures custom domain names for the API Gateway using the retrieved SSL certificates.

> To begin our customization, we will start adding configuration to this module.

Reviewer: ```suggestion
To begin our customization, we will start adding configuration to the existing module.
```

## Configure throttle limits

> Update the `aws_apigatewayv2_stage.stage" in aws/deploytf/.nitric/modules/bucket/main.tf to add default_route_settings:

Reviewer: ```suggestion
Update `aws_apigatewayv2_stage.stage` in `aws/deploytf/.nitric/modules/bucket/main.tf` to add `default_route_settings` which include throttling limits:
```

> ```hcl

Reviewer: ```suggestion
```hcl
```
```suggestion
```hcl {{ tag: 'aws/deploytf/.nitric/modules/bucket/main.tf' }}
```

resource "aws_apigatewayv2_stage" "stage" {
  api_id      = aws_apigatewayv2_api.api_gateway.id
  name        = "$default"
  auto_deploy = true

  default_route_settings {
    throttling_burst_limit = 1000
    throttling_rate_limit  = 500
  }
}
```

<Note>
  Full documentation can be found on the [Terraform
  registry](https://registry.terraform.io/providers/hashicorp/aws/2.70.2/docs/resources/apigatewayv2_stage#throttling_burst_limit).
</Note>

## Building and using your updated provider

> The Nitric project includes a make file that will build and install your provider as nitric/awstf@0.0.1 by default.

Reviewer: ```suggestion
The Nitric project includes a make file that will build and install your provider as `nitric/awstf@0.0.1` by default.
```

> Run `make install` from the provider's root directory - `nitric/cloud/aws`

Reviewer: ```suggestion
Navigate to `nitric/cloud/aws` and run `make install` to build and install the modified provider binary.

```bash
cd nitric/cloud/aws

make install
```
```

The provider can then be used directly in your project's stack file as follows.

```yaml
# The nitric provider to use
provider: nitric/awstf@0.0.1

# The target aws region to deploy to
region: us-east-2
```

<Note>
  If you don't have a stack file use the `nitric stack new` to create one.
</Note>

You'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:

> ```

Reviewer: ```suggestion
```yaml {{ tag: 'nitric.yaml' }}
```

preview:
  - beta-providers
```

> You can generate the Terraform code by running the following command:

Reviewer: ```suggestion
You can generate the Terraform project as usual by running the `nitric up` command:
```

```bash
nitric up
```

To deploy the application using Terraform, you can navigate into your Terraform stack directory and use the standard Terraform commands:

```bash
terraform init
terraform plan
terraform apply
```

You can examine your configuration by logging into the AWS console.

> <img

Reviewer 2: Is there a link to the aws console we can provide (probably on us-east-2)?

export const description =
  'Extend the Nitric AWS Terraform provider to set API Gateway throttling limits'

# Add Throttle Limits to API Gateway

Throttling helps prevent your backend services from being overwhelmed by too many requests at once. This is particularly important if your backend services have limited capacity and can only handle a certain number of requests per second.

Without throttling, a single client could potentially consume all available resources, leaving others with degraded service or no service at all.

## What we'll be doing

1. Review the existing module
2. Configure throttle limits

## Review the existing module

> Start by cloning the [Nitric repository](https://github.com/nitrictech/nitric, then examine how the Terrraform provider [provisions an API Gateway](https://github.com/nitrictech/nitric/tree/main/cloud/aws/deploytf/.nitric/modules/api).

Reviewer 3: ```suggestion
Start by cloning the [Nitric repository](https://github.com/nitrictech/nitric), then examine how the Terrraform provider [provisions an API Gateway](https://github.com/nitrictech/nitric/tree/main/cloud/aws/deploytf/.nitric/modules/api).
```

1. Defines an HTTP API Gateway with specified name, protocol, and API specification.
2. Sets up a "$default" deployment stage with automatic deployment enabled.
3. Allows the API Gateway to invoke specified Lambda functions.
4. Looks up existing certificates for specified domains.
5. Configures custom domain names for the API Gateway using the retrieved certificates.

To begin our customization, we will start adding configuration to the existing module.

## Configure throttle limits

Update `aws_apigatewayv2_stage.stage` in `aws/deploytf/.nitric/modules/bucket/main.tf` to add `default_route_settings` which include throttling limits:

```hcl {{ tag: 'aws/deploytf/.nitric/modules/bucket/main.tf' }}
resource "aws_apigatewayv2_stage" "stage" {
  api_id      = aws_apigatewayv2_api.api_gateway.id
  name        = "$default"
  auto_deploy = true

  default_route_settings {
    throttling_burst_limit = 1000
    throttling_rate_limit  = 500
  }
}
```

<Note>
  Full documentation can be found on the [Terraform
  registry](https://registry.terraform.io/providers/hashicorp/aws/2.70.2/docs/resources/apigatewayv2_stage#throttling_burst_limit).
</Note>

> ## Building and using your updated provider

Reviewer 3: ```suggestion
## Build and use your updated provider
```

The Nitric project includes a make file that will build and install your provider as `nitric/awstf@0.0.1` by default.

Navigate to `nitric/cloud/aws` and run `make install` to build and install the modified provider binary.

```bash
cd nitric/cloud/aws

make install
```

The provider can then be used directly in your project's stack file as follows.

```yaml
# The nitric provider to use
provider: nitric/awstf@0.0.1

# The target aws region to deploy to
region: us-east-2
```

<Note>

>   If you don't have a stack file use the `nitric stack new` to create one.

Reviewer 3: ```suggestion
  If you don't have a stack file use `nitric stack new` to create one.
```

</Note>

> You'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:

Reviewer 3: ```suggestion
Because the [Terraform providers](https://nitric.io/docs/reference/providers/terraform) are in preview, you'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:
```

> Because the [Terraform providers](https://nitric.io/docs/reference/providers/terraform) are in preview, you'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:

Reviewer 2: ```suggestion
Because the [Terraform providers](/reference/providers/terraform) are in preview, you'll also need to enable `beta-providers` in your Nitric project by adding the following to your project's nitric.yaml file:
```
